Src/Base/AMReX_IntegratorBase.H

```diff
@@ -0,0 +1,191 @@
+
+    static void Saxpy(T& Y, const amrex::Real a, T& X)
+    {
+        // Calculate Y += a * X using a particle-level saxpy function supplied by the particle container T
+        typedef amrex::ParIter<T::NStructReal, T::NStructInt, T::NArrayReal, T::NArrayInt> TParIter;
+        typedef amrex::Particle<T::NStructReal, T::NStructInt> ParticleType;
+
+        int lev = 0;
+        TParIter pty(Y, lev);
+        TParIter ptx(X, lev);
+
+        auto checkValid = [&]() -> bool {
+            bool pty_v = pty.isValid();
+            bool ptx_v = ptx.isValid();
+            AMREX_ASSERT(pty_v == ptx_v);
+            return pty_v && ptx_v;
+        };
+
+        auto ptIncrement = [&](){ ++pty; ++ptx; };
+
+#ifdef AMREX_USE_OMP
+#pragma omp parallel
+#endif
+        for (; checkValid(); ptIncrement())
+        {
+            const int npy  = pty.numParticles();
+            const int npx  = ptx.numParticles();
+            AMREX_ASSERT(npy == npx);
+
+            ParticleType* psy = &(pty.GetArrayOfStructs()[0]);
+            ParticleType* psx = &(ptx.GetArrayOfStructs()[0]);
+
+            auto particle_apply_rhs = T::particle_apply_rhs;
+
+            amrex::ParallelFor ( npy, [=] AMREX_GPU_DEVICE (int i) {
+                ParticleType& py = psy[i];
+                const ParticleType& px = psx[i];
+                particle_apply_rhs(py, a, px);
+            });
+        }
+    }
+
+};
+#endif
+
+template<class T>
+struct IntegratorOps<T, typename std::enable_if<std::is_same<amrex::Vector<std::unique_ptr<amrex::MultiFab> >, T>::value>::type>
+{
+
+    static void CreateLike(amrex::Vector<std::unique_ptr<T> >& V, const T& Other, bool Grow = true)
+    {
+        // Emplace a new T in V with the same size as Other
+        V.emplace_back(std::make_unique<T>());
+        for (auto const& other_mf : Other) {
+            int nGrow = Grow ? other_mf->nGrow() : 0;
+            V.back()->emplace_back(std::make_unique<amrex::MultiFab>(other_mf->boxArray(), other_mf->DistributionMap(), other_mf->nComp(), nGrow));
+        }
+    }
+
+    static void Copy(T& Y, const T& Other, bool Grow = true)
+    {
+        // Copy the contents of Other into Y
+        const int size = Y.size();
+        for (int i = 0; i < size; ++i) {
+            int nGrow = Grow ? Other[i]->nGrow() : 0;
+            amrex::MultiFab::Copy(*Y[i], *Other[i], 0, 0, Other[i]->nComp(), nGrow);
+        }
+    }
+
+    static void Saxpy(T& Y, const amrex::Real a, const T& X, bool Grow = false)
+    {
+        // Calculate Y += a * X
+        const int size = Y.size();
+        for (int i = 0; i < size; ++i) {
+            int nGrow = Grow ? X[i]->nGrow() : 0;
+            amrex::MultiFab::Saxpy(*Y[i], a, *X[i], 0, 0, X[i]->nComp(), nGrow);
+        }
+    }
+
+};
+
+template<class T>
+struct IntegratorOps<T, typename std::enable_if<std::is_same<amrex::MultiFab, T>::value>::type>
+{
+
+    static void CreateLike(amrex::Vector<std::unique_ptr<T> >& V, const T& Other, bool Grow = false)
```

Comment:
Does it make sense to make `Grow` a required argument?  The default parameter is false here, but true for other variants.  That might be confusing.

Reply:
Ah yes, that does make sense, I'll have to make some changes along with that ...

Reply:
I've incorporated all the other suggested changes. Coming back to this, there is a reason for the way the default arguments are set up here.

It has to do with how these `IntegratorOps` functions are used inside the integrators:

* `CreateLike` is only used to create `T` objects to hold right-hand side evaluations, so it will never need to store ghost cells.
* `Copy` is used to copy state or solution data from one `T` object to another. They could be old-time data or Runge-Kutta stage data. Either way, that copy always needs to include ghost cells to allow right-hand side evaluations.
* `Saxpy` only affects valid data, not ghost data, because we never integrate the ghost cells themselves.

So these choices for the `Grow` argument should be false in all cases except `Copy` just because of how the integrator classes use these functions. But I can't anticipate how future code might want to use these functions, so I don't want to hard-code these defaults if I can help it, because hiding that behavior might lead to bugs in the future if someone assumes ghost cells are always included or not. On the other hand, these arguments will always be the same inside the integrator classes, so it seems cleaner not to pass an extra argument that will always be the same for each function.

Another reason I didn't make them required arguments is that they have no meaning for the case that `T` is a `ParticleContainer` since I didn't implement anything to support a particle container having ghost particles. It's not a use case I intended.

If it would be better to make these required arguments anyway, I'm happy to do it. Just explaining why I wrote things this way initially.